miniconf/src/node.rs

```diff
@@ -0,0 +1,263 @@
+    fn try_from(value: Result<usize, Error<()>>) -> Result<Self, Traversal> {
+        match value {
+            Ok(depth) => Ok(Node::leaf(depth)),
+            Err(Error::Traversal(Traversal::TooShort(depth))) => Ok(Node::internal(depth)),
+            Err(Error::Inner(depth, ())) => Err(Traversal::TooShort(depth)),
+            Err(Error::Traversal(err)) => Err(err),
+            Err(Error::Finalization(())) => unreachable!(),
+        }
+    }
+}
+
+/// Look up an `IntoKeys` in a `TreeKey` and transcode into `Self`.
+pub trait Transcode {
+    /// Perform a lookup and transcode the keys into self
+    ///
+    /// Returns node type and depth info.
+    ///
+    /// `Err(Traversal::TooShort(depth))` indicates that `Self` does not have sufficient
+    /// capacity and failed to encode the key at the given depth.
+    fn transcode<M, const Y: usize, K>(&mut self, keys: K) -> Result<Node, Traversal>
+    where
+        M: TreeKey<Y> + ?Sized,
+        K: IntoKeys;
+}
+
+/// Shim to provide the bare `Node` lookup without transcoding target
+impl Transcode for () {
+    fn transcode<M, const Y: usize, K>(&mut self, keys: K) -> Result<Node, Traversal>
+    where
+        M: TreeKey<Y> + ?Sized,
+        K: IntoKeys,
+    {
+        M::traverse_by_key(keys.into_keys(), |_index, _name, _len| Ok::<_, ()>(())).try_into()
+    }
+}
+
+/// Path with named keys separated by a separator char
+///
+/// The path will either be empty or start with the separator.
+///
+/// * `path: T`: A `Write` to write the separators and node names into during `Transcode`.
+///   See also [TreeKey::metadata()] and `Metadata::max_length()` for upper bounds
+///   on path length. Can also be a `AsRef<str>` to implement `IntoKeys` (see [`KeysIter`]).
+/// * `const S: char`: The path hierarchy separator to be inserted before each name,
+///   e.g. `'/'`.
+#[derive(Debug, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+#[repr(transparent)]
+#[serde(transparent)]
+pub struct Path<T: ?Sized, const S: char>(pub T);
+
+impl<T: ?Sized, const S: char> Path<T, S> {
+    /// The path hierarchy separator
+    pub const fn separator(&self) -> char {
+        S
+    }
+}
+
+impl<T, const S: char> Path<T, S> {
+    /// Extract just the path
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+impl<T: ?Sized, const S: char> Deref for Path<T, S> {
+    type Target = T;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<T: ?Sized, const S: char> DerefMut for Path<T, S> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<T, const S: char> From<T> for Path<T, S> {
+    fn from(value: T) -> Self {
+        Path(value)
+    }
+}
+
+impl<'a, T: AsRef<str> + ?Sized, const S: char> IntoKeys for &'a Path<T, S> {
+    type IntoKeys = KeysIter<Skip<Split<'a, char>>>;
+    fn into_keys(self) -> Self::IntoKeys {
+        self.0.as_ref().split(self.separator()).skip(1).into_keys()
```

Comment:
Doesn't this assume that the path begins with the separator? If the path didn't, wouldn't we skip the root element?

Reply:
The rationale is as before. `Path`s start with the separator (as documented). The issue is `"".split('/') -> [""]`: disambiguate between e.g. `Option<T>: TreeKey<1>` which does not consume a key and the empty name, which consumes a key.